pm_tb_data/fetch/nrt.py

```diff
@@ -0,0 +1,186 @@
+        s = requests.session()
+
+    for host in hosts:
+        resp = s.get(
+            host,
+            # We only want to inspect the redirect, not follow it yet:
+            allow_redirects=False,
+            # We don't want to accidentally fetch any data:
+            stream=True,
+            verify=verify,
+        )
+        # Copy the headers so they can be used case-insensitively after the
+        # response is closed.
+        headers = {k.lower(): v for k, v in resp.headers.items()}
+        resp.close()
+
+        redirected = resp.status_code == 302
+        redirected_to_urs = (
+            redirected and "urs.earthdata.nasa.gov" in headers["location"]
+        )
+
+        if not (redirected_to_urs):
+            print(f"Host {host} did not redirect to URS -- continuing without auth.")
+            return s
+
+        auth_resp = s.get(
+            headers["location"],
+            # Don't download data!
+            stream=True,
+            auth=_get_earthdata_creds(),
+        )
+        resp.close()
+        if not (auth_resp.ok and s.cookies.get(_URS_COOKIE) == "yes"):
+            msg = f"Authentication with Earthdata Login failed with:\n{auth_resp.text}"
+            raise RuntimeError(msg)
+
+        print(f"Authenticated for {host} with Earthdata Login.")
+
+    return s
+
+
+FileType = Literal["R04", "P04"]
+
+
+class GranuleInfo(TypedDict):
+    file_type: FileType
+    filename: str
+    data_url: str
+
+
+GranuleInfoByDate = dict[dt.date, GranuleInfo]
+
+
+def _get_granule_info_by_date(*, data_granules: list[DataGranule]) -> GranuleInfoByDate:
+    fn_pattern = re.compile(
+        r"AMSR_U2_L3_SeaIce12km_(?P<file_type>P04|R04)_(?P<file_date>\d{8}).he5"
+    )
+    # TODO: better name/data structure.
+    granules_by_date: GranuleInfoByDate = {}
+    for granule in data_granules:
+        # The `native-id` of each granule is the filename. E.g.,
+        # `AMSR_U2_L3_SeaIce12km_R04_20230930.he5`.
+        filename = granule["meta"]["native-id"]
+        if not (match := fn_pattern.match(filename)):
+            # TODO: custom `pm_tb_data` error
+            raise FetchRemoteDataError(
+                "Found unexpected filename in CMR results (`native-id`)"
+                f": {filename}."
+            )
+        file_type = match.group("file_type")
+        file_type = cast(FileType, file_type)
+        file_date_str = match.group("file_date")
+        file_date = dt.datetime.strptime(file_date_str, "%Y%m%d").date()
+        # There are two links for each granule. one for lance.nsstc.nasa.gov and
+        # the other for lance.itsc.uah.edu. The first one is fine.
+        data_url = granule.data_links(access="external")[0]
+        granules_by_date[file_date] = {
+            "file_type": file_type,
+            "filename": filename,
+            "data_url": data_url,
+        }
+
+    return granules_by_date
+
+
+def _filter_out_last_day(*, granules_by_date: GranuleInfoByDate) -> GranuleInfoByDate:
```

Comment:
Filter out the last day of available data, unless it's an `*_R04_*` file.

We don't consider LANCE AMSR2 data to be ready for NRT use in e.g., the `seaice_ecdr` until the file is marked complete (`*_R04_*`) or a more recent data file is available. The latest available `*_P04_*` file is subject to change, so we exclude it.